opm/simulators/linalg/bda/WellContributions.hpp

```diff
@@ -71,30 +71,44 @@ class WellContributions
         B
     };
 
-    unsigned int dim;                        // number of columns in blocks in B and C, equal to StandardWell::numEq
-    unsigned int dim_wells;                  // number of rows in blocks in B and C, equal to StandardWell::numStaticWellEq
-    std::vector<MultisegmentWellContribution*> multisegments;
-
-#if HAVE_OPENCL
-    std::vector<double> h_Cnnzs_ocl, h_Dnnzs_ocl, h_Bnnzs_ocl;
-    std::vector<int> h_Ccols_ocl, h_Bcols_ocl;
-    std::vector<unsigned int> h_val_pointers_ocl;
-#endif
-
 private:
     bool opencl_gpu = false;
     bool cuda_gpu = false;
+    bool allocated = false;
 
     unsigned int N;                          // number of rows (not blockrows) in vectors x and y
-    unsigned int num_ms_wells = 0;           // number of MultisegmentWells in this object, must equal multisegments.size()
-
-#if HAVE_CUDA
-    bool allocated = false;
+    unsigned int Nb;                         // number of blockrows in vectors x and y
+    unsigned int dim;                        // number of columns in blocks in B and C, equal to StandardWell::numEq
+    unsigned int dim_wells;                  // number of rows in blocks in B and C, equal to StandardWell::numStaticWellEq
     unsigned int num_blocks = 0;             // total number of blocks in all wells
     unsigned int num_std_wells = 0;          // number of StandardWells in this object
+    unsigned int num_ms_wells = 0;           // number of MultisegmentWells in this object, must equal multisegments.size()
     unsigned int num_blocks_so_far = 0;      // keep track of where next data is written
     unsigned int num_std_wells_so_far = 0;   // keep track of where next data is written
     unsigned int *val_pointers = nullptr;    // val_pointers[wellID] == index of first block for this well in Ccols and Bcols
+
+    double *h_x = nullptr;
+    double *h_y = nullptr;
+    std::vector<MultisegmentWellContribution*> multisegments;
+
+#if HAVE_OPENCL
+    typedef cl::make_kernel<cl::Buffer&, cl::Buffer&, cl::Buffer&, cl::Buffer&, cl::Buffer&, cl::Buffer&, cl::Buffer&,
+                            cl::Buffer&, const unsigned int, const unsigned int, cl::Buffer&,
+                            cl::LocalSpaceArg, cl::LocalSpaceArg, cl::LocalSpaceArg> kernel_type;
+
+    cl::Context *context;
+    cl::CommandQueue *queue;
+    kernel_type *kernel;
+
+    std::unique_ptr<cl::Buffer> d_Cnnzs_ocl, d_Dnnzs_ocl, d_Bnnzs_ocl;
```

Comment:
Do we need the `unique_ptr`? `clBuffer` seems to be capable of managing the memory (default constructor uses NULL an later one can do `d_Cnnzs_ocl = clBuffer(...)`. But there might a downside that I do not see.